aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -727,5 +727,214 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
+  THGenerator* generator = get_generator(nullptr);
+
+  if (weights.numel() == 0){ // Uniform Sapling
+    Tensor indices_n = at::arange({n}, options);
+
+    // This is a trick to speed up the reservoir sampling.
+    // It makes the worst case be k = n / 2.
+    int split, begin, end;
+    if(2 * k < n){
+      split = n - k;
+      begin = n - k;
+      end = n;
+    } else {
+      split = k;
+      begin = 0;
+      end = k;
+    }
+
+  reservoir_generator_cpu(
+    indices_n.data<int64_t>(),
+    n,
+    split,
+    generator
+	);
+
+  return x.index_select(
+    0,
+    indices_n.index_select(
+      0,
+      at::arange(begin, end, options)
+    )
+  );
+
+} else { // Weighted Sapling
+
+		// If the weights are contiguous floating points, then
+		// the next step won't generate a copy.
+	  Tensor weights_contiguous = weights.contiguous().to(at::kFloat);
+
+	  AT_CHECK(
+	    weights_contiguous.device() == x.device(),
+	    "The weights must share the same device as the inputs."
+	  );
+
+	  AT_CHECK(
+	    n == weights_contiguous.numel(),
+	    "The weights must have the same number of elements as the input's first dimension."
+	  );
+
+	  AT_CHECK(
+	    weights_contiguous.dim() == 1,
+	    "The weights must 1-dimensional."
+	  );
+
+	  AT_CHECK(
+	    weights_contiguous.nonzero().numel() >= k,
+	    "Cannot have less non-zero weights than the number of samples."
+	  );
+
+	  AT_CHECK(
+	    weights_contiguous.min().item().toLong() >= 0,
+	    "All the weights must be non-negative."
+	  );
+
+	  Tensor keys = at::empty({n}, weights_contiguous.options());
+
+    generate_keys(
+      keys.data<float>(),
+      weights_contiguous.data<float>(),
+      n,
+      generator);
+
+	  return x.index_select(0, std::get<1>(keys.topk(k)));
+  }
+}
+
+Tensor sampling_with_replacement_cpu(
+  const Tensor& x,
+  const Tensor& weights,
+  int64_t k
+){
+
+	AT_CHECK(
+    x.dim() > 0,
+    "The input Tensor must have at least one dimension"
+  );
+
+  int n = x.size(0);
+  Tensor samples;
+
+  if (weights.numel() == 0){ // Uniform Sapling
+    samples = at::randint(0, n, {k}, x.options().dtype(at::kLong));
+  } else { // Weighted Sapling
+
+    AT_CHECK(
+      weights.min().item().toLong() >= 0,
+      "All the weights must be non-negative."
+    );
+
+    AT_CHECK(
+      weights.max().item().toLong() > 0,
+      "The sum of all the weights must be strictly greater than zero."
+    );
+
+    AT_CHECK(
+	    n == weights.numel(),
+	    "The weights must have the same number of elements as the input's first dimension."
+	  );
+
+	  AT_CHECK(
+	    weights.dim() == 1,
+	    "The weights must 1-dimensional."
+	  );
+
+    THGenerator* generator = get_generator(nullptr);
+    samples = at::empty({k}, x.options().dtype(at::kLong));
+    int64_t *samples_ptr = samples.data<int64_t>();
+
+    Tensor cdf = weights.cumsum(0).to(at::kFloat);
+    cdf /= cdf[-1];
+
+    float *cdf_ptr = cdf.data<float>();
+    for(int i = 0; i < k; i++){
```

Comment:
Do you expect people to be calling this with `k` large enough such that we would want to parallelize this part?

Reply:
It would be nice to parallelize this only when k is big enough. I was planning to perform some tests with `at::parallel_for` later, what do you think ?